### PR TITLE
Add support for Python 3.9

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Python 3.9.0 was released on 10-05-2020, this commits marks the support
of Python 3.9 in qiskit-terra. It adds the supported python version in
the package metadata and updates the CI configuration to run test jobs
on Python 3.9 and build python 3.9 wheels.

### Details and comments

Missing Support/Wheels:

requirements:
- [x] retworkx (https://github.com/Qiskit/retworkx/pull/168 ) 0.6.0
- [x] numpy (added in 1.19.3)
- [x] scipy 1.5.4

optional requirements:
- [x] tweedledum (boschmitt/tweedledum#88)
- [x] matplotlib
- [x] pandas 1.1.4
- [x] tornado 6.1

test requirements:
- [x] lazy-object-proxy 1.5.2
- [x] PyYAML 5.3.1

### CODE REVIEW

**Comment by mtreinish** [comment]
The linux Python 3.9 CI job finally passed. But none of the packages listed in the PR summary have wheels yet so they were all compiled from source for that job. A subset of those packages are also why windows and macOS jobs fail because the compilation fails in those environments. This will likely be blocked until we have at least the requirements and optional requirements updated with python 3.9 wheels

**Comment by cclauss** [comment]
Are there more dependencies with wheels now?

**Comment by mtreinish** [comment]
> Are there more dependencies with wheels now?

It looks like pandas and tornado did. I've updated the PR summary to mark that. But based on the latest run of CI here from yesterday things are still blocked on scipy not supporting python 3.9 yet. The rest of the wheels that are missing are listed after it in the logs. For example: https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=20937&view=logs&j=0966c229-dc5e-59db-1707-4394fcb5748f&t=949aafd6-37aa-5146-914f-bcc53e2bf912&l=1159 although in those cases they built successfully from source. But before we merge this we'll want all of those (in the requirements and optional requirements) to have wheels because otherwise we're forcing our users to have compilers setup locally which they often don't.

File: releasenotes/notes/python39-617f3ed823e63cfe.yaml
Comment (kdk) [inline releasenotes/notes/python39-617f3ed823e63cfe.yaml:5]: Can you add a comment that, at time of release, users who want to use matplotlib will need to build it from source? Then I think this is good to go.

Comment (mtreinish) [reply releasenotes/notes/python39-617f3ed823e63cfe.yaml:5]: Heh, I just started to write this. But it looks like there was a 3.3.3 release ~12 hours ago that added 3.9 wheels: https://pypi.org/project/matplotlib/#files

**Review by kdk — APPROVED** [review]

**Comment by kdk** [comment]
Looks like the windows tests are reliably failing for `test.python.tools.jupyter.test_notebooks.TestJupyter.test_jupyter_jobs_pbars` due to an asyncio method not implemented on windows. (ref: https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=21667&view=logs&j=7ac14b15-9bb5-559c-7b14-675b8058d5d8&t=8c2cbfe9-8c63-538d-2158-ae270fcfde5d&l=7234 )

From https://docs.python.org/3/library/asyncio-platforms.html , it looks like the default event loop changed (in python 3.8) to `ProactorEventLoop` which does not support `add_reader`.

We have a skip for this test for python 3.8 that could be updated https://github.com/Qiskit/qiskit-terra/blob/01ded73/test/python/tools/jupyter/test_notebooks.py#L73 (and to include that the osx 3.9 test must have passed).

**Comment by cclauss** [comment]
Would it be possible/better to set a different event loop that is more compatible with all platforms?  Or at least detect the OS and test on the best event loop for that OS?

**Comment by mtreinish** [comment]
> Would it be possible/better to set a different event loop that is more compatible with all platforms? Or at least detect the OS and test on the best event loop for that OS?

I don't think it's really an issue for here and is outside the scope of what that test is verifying. The event loop error is coming from inside of jupyter/nbclient and is an internal issue for jupyter. For testing we're only trying to verify that the progress bar widget works as expected. The windows test failure because of a jupyter issue isn't something we should block this on. If you have a way to fix the test so it runs on windows please push a follow up PR to do that and remove the test skip.

File: test/python/tools/jupyter/test_notebooks.py
Comment (cclauss) [inline test/python/tools/jupyter/test_notebooks.py:74]:
```suggestion
        sys.version_info >= (3, 8) and
```

**Review by cclauss — APPROVED** [review]

File: test/python/tools/jupyter/test_notebooks.py
Comment (cclauss) [inline test/python/tools/jupyter/test_notebooks.py:75]:
```suggestion
        sys.platform != 'linux', 'Fails with Python >= 3.8 on osx and windows')
```
;-)

**Review by cclauss — APPROVED** [review]

**Review by kdk — APPROVED** [review]

**Pull request merged**
